Histogram Optimized Tree Grower

hcho3 opened the pull request:
* Add a brand new updater to support histogram-based algorithm, which buckets continuous features into discrete bins to speed up training. To use it, set `tree_method=hist` to configuration.
* Support multiple tree growing strategies. For now, two policies are supported:
  * `grow_policy=depthwise` (default):  favor splitting at nodes closest to the root, i.e. grow depth-wise.
  * `grow_policy=lossguide`: favor splitting at nodes with highest loss change
* Improve single-threaded performance
  * Unroll critical loops
  * Introduce specialized code for dense data (i.e. no missing values)
* Additional training parameters: `max_leaves`, `max_bin`, `grow_policy`, `verbose`

tqchen commented:
please elaborate a bit on how each additional parameter can be used (max_leaves, max_bins, grow_policy)

hcho3 commented:
@tqchen Should I add a separate `README` for the new parameters?

tqchen commented:
no, just paste the comments in the PR to give it more context to others

hcho3 commented:
Working on fixes to pass CI...

hcho3 commented:
CI should pass now...

tqchen commented:
please fix my last comment, go through the code again and I will merge this PR in. Thanks!

hcho3 commented:
Opened an issue to document features: #1950

tqchen commented:
To make the feature available in R, add the file to 

https://github.com/dmlc/xgboost/blob/master/amalgamation/xgboost-all0.cc

tqchen commented on src/common/row_set.h:
```diff
@@ -0,0 +1,103 @@
+    bst_uint* begin = all_begin + (e.begin - all_begin);
+
+    bst_uint* it = begin;
+    for (bst_omp_uint tid = 0; tid < nthread; ++tid) {
```
add a comment: can be parallelized

hcho3 replied:
done

tqchen commented on src/tree/updater_fast_hist.cc:
```diff
@@ -0,0 +1,702 @@
+    SplitEntry best;
+    // constructor
+    explicit NodeEntry(const TrainParam& param)
+        : stats(param), root_gain(0.0f), weight(0.0f){
```
space between ```){```

hcho3 replied:
fixed

tqchen commented on src/tree/updater_fast_hist.cc:
```diff
@@ -0,0 +1,702 @@
+
+      // set all the rest expanding nodes to leaf
+      while (!qexpand_->empty()) {
+        const int nid = qexpand_->top().nid;
```
add comment:
This post condition is not needed in current code, but may be necessary when there are stopping rule that leaves qexpand non-empty

tqchen commented on src/tree/updater_fast_hist.cc:
```diff
@@ -0,0 +1,702 @@
+      int num_leaves = 0;
+
+      double tstart;
+      double init_data = 0;
```
variable name better inform what it is, change to ```time_init_data```, etc

hcho3 replied:
fixed

tqchen commented on src/tree/updater_fast_hist.cc:
```diff
@@ -0,0 +1,702 @@
+        const auto nnz = info.num_nonzero;
+        // number of discrete bins for feature 0
+        const unsigned nbins_f0 = gmat.cut->row_ptr[1] - gmat.cut->row_ptr[0];
+        if (nrow*ncol == nnz) {
```
space between operands in multiplication

hcho3 replied:
fixed

tqchen commented on src/tree/updater_fast_hist.cc:
```diff
@@ -0,0 +1,702 @@
+        if (nrow*ncol == nnz) {
+          // dense data with zero-based indexing
+          data_layout_ = kDenseDataZeroBased;
+        } else if (nbins_f0 == 0 && nrow*(ncol-1) == nnz) {
```
space between operands

hcho3 replied:
fixed

tqchen commented on src/tree/updater_fast_hist.cc:
```diff
@@ -0,0 +1,702 @@
+      }
+      if (data_layout_ == kDenseDataZeroBased || data_layout_ == kDenseDataOneBased) {
+        const std::vector<unsigned>& row_ptr = gmat.cut->row_ptr;
+        const size_t nfeature = row_ptr.size()-1;
```
space between operands

hcho3 replied:
fixed

tqchen commented on src/tree/updater_fast_hist.cc:
```diff
@@ -0,0 +1,702 @@
+        const size_t nfeature = row_ptr.size()-1;
+        size_t min_nbins_per_feature = 0;
+        for (size_t i = 0; i < nfeature; ++i) {
+          const unsigned nbins = row_ptr[i+1] - row_ptr[i];
```
space between operands

hcho3 replied:
fixed

tqchen commented on src/tree/updater_fast_hist.cc:
```diff
@@ -0,0 +1,702 @@
+      const auto& rowset = row_set_collection_[nid];
+      if (data_layout_ == kDenseDataZeroBased || data_layout_ == kDenseDataOneBased) {
+        /* specialized code for dense data */
+        const size_t column_offset = (data_layout_ == kDenseDataOneBased) ? (fid-1): fid;
```
space netween operands

hcho3 replied:
fixed

tqchen commented on src/tree/updater_fast_hist.cc:
```diff
@@ -0,0 +1,702 @@
+      ExpandEntry(int nid, int depth, bst_float loss_chg)
+        : nid(nid), depth(depth), loss_chg(loss_chg) {}
+    };
+    inline static bool depth_wise(ExpandEntry lhs, ExpandEntry rhs) {
```
The priority queue uses heap, which is in-stable, how do we break the ties? Add a timestamp of insertsion, when depth equals each other, select smaller timestamp

hcho3 replied:
I added timestamp to ExpandEntry to support tie breaking

khotilov commented:
That's a nice re-use of the existing quantile utilities!

However, while giving it a brief try, it seems to be working with sparse data, but produces a trivial model as soon as I convert the data to dense matrices:
```
> data(agaricus.train, package='xgboost')
> data(agaricus.test, package='xgboost')
> 
> dtrain <- xgb.DMatrix(as.matrix(agaricus.train$data), label = agaricus.train$label)
> dtest <- xgb.DMatrix(as.matrix(agaricus.test$data), label = agaricus.test$label)
> watchlist <- list(eval = dtest, train = dtrain)
> 
> param <- list(max_depth = 2, eta = 1, silent = 1, nthread = 2, 
+               objective = "binary:logistic", eval_metric = "auc",
+               tree_method="hist", max_bin = 256, verbose=1)
> bst <- xgb.train(param, dtrain, nrounds = 2, watchlist)
[13:54:53] Tree method is selected to be 'hist', which uses histogram aggregation for faster training. Using default sequence of updaters: grow_fast_histmaker,prune
[13:54:53] amalgamation/../src/tree/updater_fast_hist.cc:56: Generating gmat: 0.0140374 sec
[13:54:54] amalgamation/../src/tree/updater_fast_hist.cc:227: 
InitData:          0.00 (50.00%)
InitNewNode:       0.00 ( 0.00%)
BuildHist:          0.00 (50.00%)
EvaluateSplit:     0.00 ( 0.00%)
ApplySplit:        0.00 ( 0.00%)
========================================
Total:             0.00
[1]	eval-auc:0.500000	train-auc:0.500000 
[13:54:54] amalgamation/../src/tree/updater_fast_hist.cc:227: 
InitData:          0.00 (  nan%)
InitNewNode:       0.00 (  nan%)
BuildHist:          0.00 (  nan%)
EvaluateSplit:     0.00 (  nan%)
ApplySplit:        0.00 (  nan%)
========================================
Total:             0.00
[2]	eval-auc:0.500000	train-auc:0.500000 
> xgb.dump(bst)
[1] "booster[0]"           "0:leaf=-0.0715053"    "booster[1]"           "0:leaf=-7.42913e-005"
```

hcho3 commented:
@khotilov Thanks a lot for the report. We are working hard to resolve this issue.

khotilov commented:
@hcho3 Thanks! The last commit does fix the problem.

I have a question about the need to pad the cut values by `rt_eps = 1e-6f`. This constant was primarily used to compare loss values; but here you are re-purposing it and are subtracting it from absolute values of features in a model. This makes things dependent on feature scales (which is unusual condition for tree models):

- obviously, one cannot have features that are on small absolute scales comparable to rt_eps
- for features with large values, subtracting rt_eps would make no numerical effect - which is like having no rt_eps at all.

What was the reason for this epsilon-padding? And if it is really needed, would it make sense to make it scale-adaptive?

hcho3 commented:
@khotilov You're right in that padding would be problematic for "small" continuous features. We removed the padding from the code.
The original reason for the padding was to make sure that categorial/ordinal feature values fall into correct bins. We created a special code to deal with this scenario.

khotilov commented:
To clarify: while the rt_eps is also used in colmaker in + or - combination with feature values (not just for loss change, as I remembered it earlier), it is used in there in such ways that it still works with small-valued features that I was too lazy to transform. But things break in fast_histmaker when bins are padded by rt_eps that is relatively large to feature scale.

And here's one more edge case that breaks fast_histmaker:
```R
library(data.table)
library(xgboost)

# simulate binary outcome dependent on one feature
N <- 10000
Scale <- 1 #0.00000001
set.seed(1)
#X <- data.table(x = 1.*rbinom(N,1,0.05))
X <- data.table(x = sample(-1:1, N, replace = T, prob = c(0.2,0.6,0.2))  )
pr = X[, plogis(-3 + 2.5 * x^2)]
y = rbinom(N, 1, pr) 
X[, x := Scale * x]
plot(factor(y) ~ factor(X$x))
#X[sample.int(N,100), x:=NA]

dtrain <- xgb.DMatrix(as.matrix(X[itr]), label=y[itr])

param <- list(objective="binary:logistic", eval_metric="auc", nthread=4, 
              max_depth=1, eta = 0.5, base_score=mean(getinfo(dtrain, "label")),
              tree_method="hist", max_bin = 2, verbose = 1)
bst <- xgb.train(param, dtrain, 5)
```
this results in
![image](https://cloud.githubusercontent.com/assets/1613551/21916093/bc8bfb2e-d903-11e6-93a6-3ad9a1bb2ac7.png)
```
[20:12:50] Tree method is selected to be 'hist', which uses histogram aggregation for faster training. Using default sequence of updaters: grow_fast_histmaker,prune
[20:12:50] amalgamation/../src/tree/../common/quantile.h:384:  check quantile stats, nbig=0, n=1
[20:12:50] amalgamation/../src/tree/../common/quantile.h:385:  srcsize=3, maxsize=2, range=3050, chunk=6100
[20:12:50] amalgamation/../src/tree/../common/quantile.h:284: [0] rmin=0, rmax=1049, wmin=1049, v=-1
[20:12:50] amalgamation/../src/tree/../common/quantile.h:284: [1] rmin=1049, rmax=4099, wmin=3050, v=0
[20:12:50] amalgamation/../src/tree/../common/quantile.h:284: [2] rmin=4099, rmax=5000, wmin=901, v=1
Error in xgb.iter.update(bst$handle, dtrain, iteration - 1, obj) : 
  [20:12:50] amalgamation/../src/tree/../common/quantile.h:388: Check failed: nbig < n - 1 quantile: too many large chunk
```

khotilov commented:
@hcho3 I've posted my last comment before I saw yours. The last updates makes it possible to use small-scale features. Thanks! 

The `max_bins=2` error is still there though.

tqchen commented:
@khotilov It appears that you hit a corner case of quantile sketch. 

Can you see if https://github.com/dmlc/xgboost/pull/1959 fixes the problem?

khotilov commented:
@tqchen Thanks, that does fix it.

hcho3 commented:
@tqchen @khotilov #1959 indeed fixes the problem with `max_bin=2`. I've incorporated the fix into this pull request. I've also added a small python test to test for `max_bin=2`.

CodingCat commented:
so, is it safe to say it is ready to add language-bindings?

tqchen commented:
yap, I think these options should be exposed.
